make PVProvider optional in server config; disallow snap/restore PVs …

skriss opened the pull request:
…when not provided

Fixes #44 

Signed-off-by: Steve Kriss <steve@heptio.com>

skriss commented:
ref. https://github.com/heptio/ark/pull/35#discussion_r132951946

ncdc commented on pkg/controller/gc_controller.go:
```diff
@@ -115,6 +115,12 @@ func (c *gcController) cleanBackups() {
 
 			for _, volumeBackup := range backup.Status.VolumeBackups {
+				if c.snapshotService == nil {
+					glog.Warningf("Cannot garbage-collect snapshot %s for backup %s/%s because server is not configured with PersistentVolumeProvider",
```
Copying my comment from the original PR:

> If the snapshot service is nil, the net result is that backups will be GC'd but associated snapshots will be orphaned?

Steve's reply:
> yes. it's a strange situation to be in - you created backups with snapshots, meaning you had a snapshotservice configured, but now you're in a mode where if you restore from them you won't use them. i suppose the alternative is to not GC any backups with snapshots in this scenario - WDYT?

ncdc replied:
I don't want to orphan data if we can avoid it

ncdc replied:
Let's put a check before the DeleteBackup call to see if the backup has any VolumeBackups. If it does and the snapshotService is nil, then log and skip deleting the backup.

skriss commented:
@ncdc still need to test and update some of the examples to remove config that's no longer necessary, but you can take a look at the code now.

ncdc commented on pkg/backup/volume_snapshot_action.go:
```diff
@@ -38,11 +39,15 @@ type volumeSnapshotAction struct {
 var _ Action = &volumeSnapshotAction{}
 
-func NewVolumeSnapshotAction(snapshotService cloudprovider.SnapshotService) Action {
+func NewVolumeSnapshotAction(snapshotService cloudprovider.SnapshotService) (Action, error) {
```
The more I think about it, I'm not sure that changing the signature here really provides much value?

skriss replied:
the intent was to embed the nil-checks into the library so if someone uses it outside the ark server they still get argument validation. what would you do differently? leave as it was?

ncdc replied:
I can see that... what I was thinking was "you're not going to call NewVolumeSnapshotAction if you don't have a snapshotService", but you never know what crazy things people will do 😄 . This is fine to keep.

ncdc commented on pkg/controller/gc_controller.go:
```diff
@@ -108,26 +108,36 @@ func (c *gcController) cleanBackups() {
-			if err := c.backupService.DeleteBackup(c.bucket, backup.Name); err != nil {
-				glog.Errorf("error deleting backup %s/%s: %v", backup.Namespace, backup.Name, err)
-			}
+		if !backup.Status.Expiration.Time.Before(now) {
```
Do you think checking for "expiration after now" is clearer than "expiration not before now"?

skriss replied:
I had the same thought, decided it was tomayto-tomahto, but if switching it strikes you as more clear I can do that.

ncdc replied:
whatevs :)

skriss commented:
@abiogenesis-now I decided it was most clear to break out the reference tables for config based on what it's being used for. PTAL.

skriss commented:
I've tested these changes on all of the supported platforms

skriss commented on docs/config-definition.md:
```diff
@@ -25,17 +25,12 @@ metadata:
   name: default
 persistentVolumeProvider:
   aws:
-    region: minio
```
Note that I switched this example to AWS because minio PVProvider config is no longer required (or valid), and unfortunately we can't currently mix/match AWS PVProvider with minio backup storage provider due to the single source of credentials.

abiogenesis-now replied:
that makes sense

skriss commented on docs/config-definition.md:
```diff
@@ -63,22 +58,44 @@ The configurable parameters are as follows:
+
+#### backupStorageProvider
+
+No parameters required; specify an empty object per [example file][13].
```
@ncdc @abiogenesis-now does this seem reasonable?

abiogenesis-now replied:
👍

abiogenesis-now commented on docs/config-definition.md:
```diff
@@ -63,22 +58,44 @@ The configurable parameters are as follows:
-| `availabilityZone` | string | Required Field | *Example*: "us-east-1a"<br><br>See [AWS documentation][4] for details. |
 | `disableSSL` | bool | `false` | Set this to `true` if you are using Minio (or another local, S3-compatible storage service) and your deployment is not secured. |
 | `s3ForcePathStyle` | bool | `false` | Set this to `true` if you are using a local storage service like Minio. |
 | `s3Url` | string | Required field for non-AWS-hosted storage| *Example*: http://minio:9000<br><br>You can specify the AWS S3 URL here for explicitness, but Ark can already generate it from `region`, `availabilityZone`, and `bucket`. This field is primarily for local sotrage services like Minio.|
```
mind fixing my typo? s/sotrage/storage
thanks!

skriss replied:
fixed.

abiogenesis-now commented on examples/azure/10-ark-config.yaml:
```diff
@@ -24,9 +24,7 @@ persistentVolumeProvider:
   bucket: <YOUR_BUCKET>
-  azure:
-    location: <YOUR_LOCATION>
-    apiTimeout: <YOUR_TIMEOUT>
```
do we not use the Azure API for object storage? or is it faster/doesn't require `apiTimeout` toggling?

skriss replied:
The slow API call issue is only with snapshot operations -- I wasn't actually using the timeout value for object storage API calls before anyway. I could keep it and switch to actually using it in the code, but I don't think it's necessary.

abiogenesis-now replied:
ah ok, I think it's fine to leave it out then!

abiogenesis-now commented on docs/config-definition.md:
```diff
@@ -63,22 +58,44 @@ The configurable parameters are as follows:
 | `s3Url` | string | Required field for non-AWS-hosted storage| *Example*: http://minio:9000<br><br>You can specify the AWS S3 URL here for explicitness, but Ark can already generate it from `region`, `availabilityZone`, and `bucket`. This field is primarily for local sotrage services like Minio.|
 | `kmsKeyID` | string | Empty | *Example*: "502b409c-4da1-419f-a16e-eif453b3i49f"<br><br>Specify an [AWS KMS key][12] id to enable encryption of the backups stored in S3. Only works with AWS S3 and may require explicitly granting key usage rights.|
 
+#### persistentVolumeProvider
```
Specify (AWS Only)? To be clearer that Minio is not supported

skriss replied:
fixed.

abiogenesis-now commented on docs/config-definition.md:
```diff
@@ -50,7 +45,7 @@ The configurable parameters are as follows:
 | Key | Type | Default | Meaning |
 | --- | --- | --- | --- |
-| `persistentVolumeProvider` | CloudProviderConfig<br><br>(Supported key values are `aws`, `gcp`, and `azure`, but only one can be present. See the corresponding [AWS][0], [GCP][1], and [Azure][2]-specific configs.) | Required Field | The specification for whichever cloud provider the cluster is using for persistent volumes (to be snapshotted).<br><br> *NOTE*: For Azure, your Kubernetes cluster n ...
+| `persistentVolumeProvider` | CloudProviderConfig<br><br>(Supported key values are `aws`, `gcp`, and `azure`, but only one can be present. See the corresponding [AWS][0], [GCP][1], and [Azure][2]-specific configs.) | None (Optional) | The specification for whichever cloud provider the cluster is using for persistent volumes (to be snapshotted), if any.<br><br>If not specified, Backups and Restor ...
```
Have we tested this in conjunction with the other PR https://github.com/heptio/ark/pull/45 that sets `--snapshot-volumes` and `--restore-volumes` to default true?

The author of that PR said it was fine to leave the quickstart commands without `--snapshot-volumes=false` since the basic Nginx example doesn't have PVs. I'm wondering if we actually do need to specify this flag now in the sample commands, since it looks it will error without a `peristentVolumeProvider` section specified. It's not a big deal, but if that's the case, we'll need to make sure to change the backup/restore commands in one of these two PRs. @ncdc does that seem right?

ncdc replied:
@skriss is the primary motivation for making PVProvider optional to support running non-cloud (i.e. minio)?

I do think @abiogenesis-now brings up a good point. As far as UX goes, I think having a backup/restore fail validation because you forgot to set `--snapshot-volumes=false` might get annoying after a while, if the server isn't configured appropriately. I guess we might need to decide how likely running without a PV provider is. Also need to decide if we need to distinguish between something like "minio and legitimately no PV provider" and "aws/gcp/azure and I forgot to configure the PV provider, so it should be a validation error". Let's continue to discuss...

skriss replied:
Hmm, yeah that is not a great UX now that true is becoming the default. We could just default to true if the PVProvider is configured, false otherwise. 

Minio is one use case; I also had in mind clusters not running in the cloud that users still want to back up to S3 (for example). This is supported but currently requires specifying a bogus PVProvider config.

ncdc replied:
> We could just default to true if the PVProvider is configured, false otherwise.

I'm liking this. If you try to take a backup and the PVProvider wasn't configured, all you get is a backup but no snapshots.

However, if a backup does have snapshots, and then Ark was reconfigured for whatever reason with no PVProvider, maybe the restore should fail validation ("backup has volume snapshots but no PRProvider configured; please contact your Ark administrator to enable the PRProvider etc etc etc")?

skriss replied:
The only thing I don't like about this is that there's no way around it - a user can't "override" the error.

ncdc replied:
You can override if you set restore snapshots to false?

skriss replied:
isn't that kind of the same problem we're trying to avoid re: UX (having to remember to specify that)? I was assuming we'd also default the flag to false for restores when no PVProvider.

skriss replied:
I'm thinking that maybe what we should do is if you're restoring a backup that has snapshots linked, and --restore-volumes=false (either because you set it or because you don't have a PV provider so we default it to false), that we put a warning in the restore operation. I'll run it by joe.

ncdc replied:
Like a prompt to the user where they have to indicate if they wish to proceed? And suppressible via `--ignore-warnings` or `--force` or something like that?

skriss replied:
talked to Joe - he recommends we make --snapshot and --restore default to "auto", which means do the sensible thing based on presence of PVProvider and/or snapshots in the backup. Explicitly setting to yes can then cause validation errors if there's no PVProvider. In the case where there are snapshots in the backup but no PVProvider for restore, we should put a warning in the Restore.

ncdc commented:
@skriss is this ready to go?

skriss commented:
Not quite - I switched to the 3-way flag and need to do some more testing after that change.

skriss commented:
@ncdc this is ready for another review. Main changes since the last time you looked:

* --snapshot-volumes and --restore-volumes have become nullable bool flags, where null means "auto" (default), which means to do the logical thing based on existence of PV Provider config
* add warning to the restore if the backup has snapshots but there's no PV Provider to restore them
* consolidate unstructured PV manipulation into `util/kube`

ncdc commented on pkg/cmd/cli/backup/create.go:
```diff
@@ -70,19 +70,29 @@ func NewCreateOptions() *CreateOptions {
 	flags.Var(&o.Labels, "labels", "labels to apply to the backup")
 	flags.VarP(&o.Selector, "selector", "l", "only back up resources matching this label selector")
+
+	// we have to construct this directly because we want to specify a NoOptDefVal to maintain
```
Why not let the default be nil?

ncdc commented on pkg/cmd/server/server.go:
```diff
@@ -242,112 +239,108 @@ func (s *server) watchConfig(config *api.Config) {
 
 func (s *server) initSnapshotService(config *api.Config) error {
+	if config.PersistentVolumeProvider == nil {
+		glog.Infof("PersistentVolumeProvider config not provided, skipping SnapshotService creation")
```
Instead of saying "skipping SnapshotService creation", I would say "volume snapshots and restores are disabled"?

ncdc commented on pkg/restore/restorers/pv_restorer.go:
```diff
@@ -43,31 +44,79 @@ func (sr *persistentVolumeRestorer) Handles(obj runtime.Unstructured, restore *a
+	var warning error
+
+	if sr.snapshotService == nil && len(backup.Status.VolumeBackups) > 0 {
+		warning = fmt.Errorf("Restore %s/%s: backup %s/%s has PV snapshots but Ark server is not configured with a PersistentVolumeProvider so cannot restore thenm", restore.Namespace, restore.Name, backup.Namespace, backup.Name)
```
Don't need to include the restore's or backup's namespace/name.

`"unable to restore PV snapshots: Ark server is not configured with a PersistentVolumeProvider"`
